### PR TITLE
EdDSA key gen improvements

### PR DESCRIPTION
**Pull request by zwilling**
Proper EdDSA key generation consistent with Go library used for EdDSA and the EdDSA specification
- Fixed length of EdDSA private key to 32 bytes
- Following EdDSA key generation specification from https://www.rfc-editor.org/rfc/rfc8032#section-5.1.5
- Managing keys as Buffer instead of strings to prevent circomlibjs and ffjavascript from reading it wrongly as utf8 input
- Integrating proper EdDSA keygen also into snap (entropy from Metamask instead of signature)

Special thanks to @HarmlessEvil for spotting these issues

### CODE REVIEW

**Review by HarmlessEvil — CHANGES_REQUESTED** [review]
Oh, no… The RFC paragraph you refering to is about how generating public from private key.

> The 32-byte **public key** is generated by the following steps.

And it is already implemented in [https://github.com/iden3/circomlibjs/blob/4f094c5be05c1f0210924a3ab204d8fd8da69f49/src/eddsa.js#L36](`priv2pub`) method of `circomlibjs`. Look, there are the steps for prunning buffer and reinterpreting array as little-endian big integer.

I must reject this pull request, because the implementation is not related to the actual generation of a private key. If you want to implement my suggestions from comments, feel free to open new pull request.

**Comment by HarmlessEvil** [comment]
I made an aditional research on whether we can slice the signature to interpret it as a random byte sequence, which is our private key. And my answer is "yes", because first 32 bytes is "r" component of signature [https://ethvigil.com/docs/eth_sign_example_code/#recovering-the-message-signer-in-the-smart-contract](proof), and this is a statistically random value as long as we treat private key as a random value [https://cryptobook.nakov.com/digital-signatures/ecdsa-sign-verify-messages#ecdsa-sign](proof).

**Comment by zwilling** [comment]
> Oh, no… The RFC paragraph you refering to is about how generating public from private key.
> 
> > The 32-byte **public key** is generated by the following steps.
> 
> And it is already implemented in [https://github.com/iden3/circomlibjs/blob/4f094c5be05c1f0210924a3ab204d8fd8da69f49/src/eddsa.js#L36](%60priv2pub%60) method of `circomlibjs`. Look, there are the steps for prunning buffer and reinterpreting array as little-endian big integer.
> 
> I must reject this pull request, because the implementation is not related to the actual generation of a private key. If you want to implement my suggestions from comments, feel free to open new pull request.

You are the best. Thanks for reading it more carefully than I did.

File: packages/zk-certificates/test/lib/testKeyManagement.ts [hunk 1]
```diff
@@ -16,17 +16,17 @@ describe('Key Management', () => {
...
-      ethers.utils.recoverAddress(
-        ethers.utils.hashMessage(eddsaKeyGenerationMessage),
-        holderEdDSAKey,
-      ),
-    ).to.equal(holder.address);
+    // test if the keys can sign and verify a message
+    const message = Buffer.from('test message');
+    const signature = eddsa.signPedersen(holderEdDSAKey, message);
+
+    expect(eddsa.verifyPedersen(message, signature, pubKey)).to.be.true;
```
[the remark this answers is not in the document]
Comment (zwilling) [reply packages/zk-certificates/test/lib/testKeyManagement.ts:29]: It should work both. But I'll change it because Poseidon is more relevant for us.

**Comment by zwilling** [comment]
Ok, I fixed the key generation and implemented your suggestions from the comment.
Please check again @HarmlessEvil

**Comment by zwilling** [comment]
I'll create a new PR. The changes are hard to view because they also include what I merged from origin/main.

**Pull request closed without merging**